// @athena/aw-api/lib-esm/lib/api.js
"use strict";
import * as request from "superagent";
var anyRequest = request;
/**
 *
 * @class ApiClass
 * @param {(string)} [domainOrOptions] - The project domain.
 */
export class ApiClass {
    constructor(domain, logger) {
        this.logger = logger;
        this.domain = "";
        this.errorHandlers = [];
        if (domain) {
            this.domain = domain;
        }
    }
    getDomain() {
        return this.domain;
    }
    addErrorHandler(handler) {
        this.errorHandlers.push(handler);
    }
    request(method, url, body, headers, queryParameters, form, reject, resolve) {
        if (this.logger) {
            this.logger.log(`Call ${method} ${url}`);
        }
        let req = (anyRequest.default || anyRequest)(method, url).query(queryParameters);
        Object.keys(headers).forEach(key => {
            req.set(key, headers[key]);
        });
        if (body) {
            req.send(body);
        }
        if (typeof (body) === 'object' && !(body.constructor.name === 'Buffer')) {
            req.set('Content-Type', 'application/json');
        }
        if (Object.keys(form).length > 0) {
            req.type('form');
            req.send(form);
        }
        req.end((error, response) => {
            if (error || !response.ok) {
                reject(error);
                this.errorHandlers.forEach(handler => handler(error));
            }
            else {
                resolve(response);
            }
        });
    }
}
export default ApiClass;
