// @athena/aw-api/lib/lib/api.js
"use strict";
import * as request from "superagent";
var anyRequest = request;
/**
 *
 * @class ApiClass
 * @param {(string)} [domainOrOptions] - The project domain.
 */
var ApiClass = /** @class */ (function () {
    function ApiClass(domain, logger) {
        this.logger = logger;
        this.domain = "";
        this.errorHandlers = [];
        if (domain) {
            this.domain = domain;
        }
    }
    ApiClass.prototype.getDomain = function () {
        return this.domain;
    };
    ApiClass.prototype.addErrorHandler = function (handler) {
        this.errorHandlers.push(handler);
    };
    ApiClass.prototype.request = function (method, url, body, headers, queryParameters, form, reject, resolve) {
        var _this = this;
        if (this.logger) {
            this.logger.log("Call " + method + " " + url);
        }
        var req = (anyRequest.default || anyRequest)(method, url).query(queryParameters);
        Object.keys(headers).forEach(function (key) {
            req.set(key, headers[key]);
        });
        if (body) {
            req.send(body);
        }
        if (typeof (body) === 'object' && !(body.constructor.name === 'Buffer')) {
            req.set('Content-Type', 'application/json');
        }
        if (Object.keys(form).length > 0) {
            req.type('form');
            req.send(form);
        }
        req.end(function (error, response) {
            if (error || !response.ok) {
                reject(error);
                _this.errorHandlers.forEach(function (handler) { return handler(error); });
            }
            else {
                resolve(response);
            }
        });
    };
    return ApiClass;
}());
export { ApiClass };
export default ApiClass;
